File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第7章砝码称重.py
#砝码称重  2021年第十一届省赛 lanqiaoOJ 1447

n = int(input())
w = list(map(int, input().split()))
ans = set()
ans.add(w[0])
for i in w[1:]:
    for j in ans.copy():
        ans.add(i)
        ans.add(j + i)
        if j - i != 0: ans.add(abs(j - i))
print(len(ans))

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第7章砝码称重.cpp
//砝码称重  2021年第十一届省赛 lanqiaoOJ 1447

#include <bits/stdc++.h>
using namespace std;
bool dp[110][200001];
int main(){
    int n; cin>>n;
    dp[0][0]=1;
    for(int i=1;i<=n;i++){
        int w; cin>>w;
        for(int j=0;j<=100000;j++)
            dp[i][j]=dp[i-1][j] | dp[i-1][abs(j-w)] | dp[i-1][j+w];
    }
    int ans = 0;
    for(int i=1;i<=100000;i++) ans += dp[n][i];
    cout<<ans;
}
